Topic System Overhaul

MCterra10 opened the pull request:
## About The Pull Request
Ike asked me really nicely so I reworked the whole topic system to use JSON and be less shit than before.
This PR ports a lot of Aurorastation/Aurora.3#825
and a little bit of Aurorastation/Aurora.3#876

### Request
Instead of passing weird arbitrary params, topic requests now require a simple JSON request:
```json
{
    "query": "api_function_key",
    "auth": "api_token",
    "source": "request source"
}
```
This is the absolute base request required for simple functions such as ping, status, playercount etc.
Many functions require additional arguments, though they're entirely reserved for internal functionality such as ahelps and crosscomms.

### Response
The server returns a simple JSON response in the following format:
```json
{
    "statuscode": 200,
    "response": "simple text information about the request's outcome",
    "data": "misc variable that can contain any additional information, such as a string or a list"
}
```
Any non-200 status code indicates an API error of some kind.

### Authentication
The authentication system has been completely overhauled in favor of a granular token permissions system.
An arbitrary amount of tokens can be created in the `comms.txt` config file, with specific permissions depending on usage.
See the config file for more elaborate configuration instructions.

***NOTE:*** By default, a static token exists which has the value of `anonymous`, and is used for unauthenticated topic calls. In the code, all topic datums have an anonymous flag, which if set to true, grants access to all tokens and the anonymous token by default.
This allows continued public access to topics such as status, playercount etc.

## Why It's Good For The Game
N/A; this changes nothing noticeable by players.

## Changelog
:cl: MCterra, Arrow768
server: Complete topic rework to use JSON and just be better in general
/:cl:

MCterra10 committed:
start the topic rework

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

MCterra10 committed:
oh yeah it's all coming together

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
noticed a bug

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
I forgot my access check

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
forgot a param requirement

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
another little bug

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
remove the insecure arg

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 commented:
Note, I still need to account for not allowing some servers access to ahelp bridging. I'll sort that out soon.

Edit: going to add a handshake protocol to authenticate trusted servers

ike709 commented:
DNM until the bot and the website are also changed to be compatible with this.

MCterra10 committed:
adds cross-server handshakes

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 commented:
alright so this is incredibly over-engineered but basically:
cross-server comms now requires a two-way "handshake" between both servers to tell each other what methods they have available and allow each server to work out what to send and what not to.

In simple terms, both servers need to have each other added as crosscomms servers, and only scopes that *they both have* are considered valid.
For example, for a server to forward ahelps to another server, it needs to verify that the other server is in possession of a key that can access ahelps on itself, and that it has a key to access ahelps on the other server.
 
Let me know if you want a terrible MS Paint flow chart of this lmao

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

MCterra10 committed:
thanks dreamchecker very cool

Crossedfall commented:
Work for this has been scheduled.

MCterra10 commented:
would be good if the code could get a pass over from a maintainer as well. I have tested it a bunch and stuff checks out, but a second pair of eyes would be good considering how core this stuff is.

MCterra10 committed:
use async proc to avoid 2 seconds additional init

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

MCterra10 commented:
Qwerty taking his time 😫

utilbird commented:
See you next year

MCterra10 commented:
😔

MCterra10 committed:
fix rate limit and auth bug

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

Crossedfall reviewed (changes requested):
Bot and website are pretty much good to go. This change needs actual documentation (i.e. a README) and conflict resolution before being ready for merge; otherwise, the major blockers are out of the way.

Can finally start going forward with this 🥳

MCterra10 committed:
Merge remote-tracking branch 'beestation/master' into topics-or-something-idk

MCterra10 commented:
Ugh I just noticed a bunch of the topic datum arg names aren't very consistent, I'll get this readme done and then fix that

MCterra10 commented:
I also wouldn't mind writing some boilerplate code for targeted topic requests and to make things a bit more universal. Right now the comms_send proc only really works for ahelps, comms consoles and newscasters.

MCterra10 committed:
general code improvements (docs, refactor, sec)
small refactor to some topic queries to modernize them and return json
adds documentation to the config readme file for topics
update topic procs to wrap world.export and handle security better

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 commented:
Code is now documented and I've made some nice improvements to the system in terms of code quality.
I'll run some tests on it tomorrow to make sure everything is in working order.

MCterra10 committed:
fix markdown formatting

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
editor issue

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Merge branch 'topics-or-something-idk' of https://github.com/MCterra10/BeeStation-Hornet into topics-or-something-idk

MCterra10 committed:
censor keys, fix bugs, url encode topic data

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 commented:
I've tested all topic endpoints (except ones that touch the DB) and made sure the responses are sane.
I've also tested cross-server comms; ahelps, comms consoles and newscasters are working.

MCterra10 committed:
update docs

Signed-off-by: Terra <terra@mcterra.id.au>

Crossedfall commented on code/datums/world_topic.dm:
```diff
@@ -1,246 +1,346 @@
+	response = versionstring
+	data = version
+
+/datum/world_topic/api_get_authed_functions
+	key = "api_get_authed_functions"
+	anonymous = TRUE
+
+/datum/world_topic/api_get_authed_functions/Run(list/input)
+	. = ..()
+	var/list/functions = GLOB.topic_tokens[input["auth"]]
+	if(functions)
+		statuscode = 200
+		response = "Authorized functions retrieved"
+		data = functions
+	else
+		statuscode = 401
+		response = "Unauthorized - No functions found"
+		data = null
```
Is it a good idea to have a function that returns the full capabilities of a token? Seems a bit like a security nightmare.

ike709 replied:
I mean I don't see what's stopping people from just attempting every API endpoint anyways.

Crossedfall replied:
Ratelimits stop them, mostly. It at least adds a significant enough delay to make it difficult to go through the entire list at once.

MCterra10 replied:
I guess I can remove/disable it if you'd like, but I really don't see how getting authorised endpoints is a security issue. If you leaked a token and someone wants to use it maliciously, removing this function will add max five minutes of work to just loop over the endpoints and sleep for 15 seconds to avoid rate limiting.

On the other hand, it's not really used for anything right now, since the only usecase within BYOND is already built into the handshake system.

All of that said, however, if you're developing an external tool, I think it's good to have a system to check what you can use, without having to actually poke the endpoints and see if you get 401 or not

Crossedfall replied:
>  removing this function will add max five minutes of work to just loop over the endpoints and sleep for 15 seconds to avoid rate limiting.

That adds a pretty obvious pattern that can be observed. Even further, it'd be trivial to setup a monitor to watch for that kind of activity which can throw an alert for us to handle.

On one hand, I can see the value of this endpoint. On the other hand, I don't particularly like that it's an incredibly obvious attack vector.

Crossedfall replied:
I am not so committed to this that I'm willing to die on the hill though. So, if the general consensus is that it'll probably be fine, I'm willing to resign from the debate and we can move forward with it still openly available.

ike709 replied:
It'll probably be fine

MCterra10 committed:
include last 4 characters of token

Signed-off-by: Terra <terra@mcterra.id.au>

Crossedfall commented:
Staging tests seem good. Pending review from another maintainer before final merge

KubeRoot commented:
This changes a lot of stuff, can somebody who knows this take a look at my PR and tell me how much was made redundant? https://github.com/BeeStation/BeeStation-Hornet/pull/4979
